### PR TITLE
Implementing predicates ordering

### LINKED ISSUE #53812: Implement predicate ordering for the kubernetes scheduler
following the design proposal on kubernetes/community#1152 this will be the issue that will track down the work to implement predicates ordering, work to be done:

* set predicate ordering to create the proposed static ordering / tests
* export restrictiveness as a prometheus metric
* write docs and examples

also to be done based on customers feedback:

* implement the end-user ordering (scheduler policy parsing + override of the static ordering) / tests

I'll create the PRs accordingly, after  kubernetes/community#1152 is merged.

@gmarek @bsalamat @k82cn

### PR DESCRIPTION
**Pull request by yastij**
**What this PR does / why we need it**: implements predicates ordering for the scheduler

**Which issue(s) this PR fixes** : Fixes #53812 

**Special notes for your reviewer**:

@bsalamat @gmarek @resouer as discussed on slack, to implement ordering we have to choices:

- use a layered approach with a list that indexes the order of the predicates map

- change the underlying data structure used to represent a collection of predicates (a map in our case) into a list of predicates objects. 
Going with this solution might be "cleaner" but it will require a lot of changes and will increase the cost for accessing predicates from O(1) to O(n) (n being the number of predicates used by the scheduler).

we might go with this solution for now. If the number of predicates start growing, we might switch to the second option.
 
**Release note**:

```release-note
adding predicates ordering for the kubernetes scheduler.
```

### CODE REVIEW

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 1]
```diff
@@ -438,34 +438,37 @@ func podFitsOnNode(
...
-			// checking other predicates?
-			if !eCacheAvailable || invalid {
-				// we need to execute predicate functions since equivalence cache does not work
-				fit, reasons, err = predicate(pod, metaToUse, nodeInfoToUse)
-				if err != nil {
-					return false, []algorithm.PredicateFailureReason{}, err
-				}
+		for _, predicateKey := range predicates.getPredicatesOrdering() {
+			//TODO (yastij) : compute average predicate restrictiveness to export it as promethus metric
+			if predicate := predicateFuncs[predicateKey]; predicate != nil {
```
Comment (bsalamat) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: You should also check if the key exists in the map.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -79,6 +79,16 @@ const (
 // For example:
 // https://github.com/kubernetes/kubernetes/blob/36a218e/plugin/pkg/scheduler/factory/factory.go#L422
 
+var (
+	predicatesOrdering = []string{"CheckNodeCondition",
```
Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: It would be great to use constants instead of string literals here.

Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: We should also add a comment that this list must be updated when new predicates are added.

**Review by bsalamat — COMMENTED** [review]
Thanks, @yastij

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 2 above)
Comment (yastij) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Yes that is the plan.

Comment (yastij) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: indeed

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (yastij) [reply plugin/pkg/scheduler/core/generic_scheduler.go]: yup i'll check the returned bool instead.

**Comment by yastij** [comment]
@bsalamat thanks for the review, updated this PR.

**Comment by dims** [comment]
/ok-to-test

@yastij i have triggered the tests ... you probably should squash the commits

**Comment by yastij** [comment]
@dims - can you add it to the 1.10 milestone ?

**Comment by dims** [comment]
@yastij i don't have enough karma to do that :) but you should be good. the milestones kick on only late in the cycle. so stuff that is not marked but that have `/lgtm` and `/approve` without any milestones will still merge in master

**Comment by yastij** [comment]
@bsalamat - some tests might fail, maybe we should add a way to append a predicate name for test purpose ?

**Comment by bsalamat** [comment]
@yastij It is fine to add a way for tests to add their predicates.

**Comment by yastij** [comment]
SGTM

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 3]
```diff
@@ -49,9 +49,25 @@ import (
...
+	PodToleratesNodeTaintsPred          = "PodToleratesNodeTaints"
+	PodToleratesNodeNoExecuteTaintsPred = "PodToleratesNodeNoExecuteTaints"
+	CheckNodeLabelPresencePred          = "CheckNodeLabelPresence"
+	checkServiceAffinityPred            = "checkServiceAffinity"
+	MaxEBSVolumeCountPred               = "MaxEBSVolumeCount"
+	MaxGCEPDVolumeCountPred             = "MaxGCEPDVolumeCount"
+	MaxAzureDiskVolumeCountPred         = "MaxAzureDiskVolumeCount"
+	NoVolumeZoneConflictPred            = "NoVolumeZoneConflict"
+	CheckNodeMemoryPressurePred         = "CheckNodeMemoryPressure"
+	CheckNodeDiskPressure               = "CheckNodeDiskPressure"
```
Comment (resouer) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Thanks for define this. Would you mind to replace all hard coded predicate names across the scheduler pkg? You can do this in a separate commit. I once added a TODO but can't find ...

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 4]
```diff
@@ -79,6 +95,19 @@ const (
 // For example:
 // https://github.com/kubernetes/kubernetes/blob/36a218e/plugin/pkg/scheduler/factory/factory.go#L422
 
+// IMPORTANT: this list contains the ordering of the predicates, if you develop a new predicates
+// it is mandatory to add its name on this list.
+// otherwise it won't be processed, see generic_scheduler#podFitsOnNode()
+var (
+	predicatesOrdering = []string{CheckNodeConditionPred,
```
Comment (resouer) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Need to explain how the order is defined in the comment.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 3 above)
Comment (yastij) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Yea sure, can you double check predicates names, afraid of having a typo (which will break existing policies)

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 4 above)
Comment (yastij) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: I'll explain briefly how the order is defined and point to design doc

Comment (resouer) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Yes, please. This is very important because we need to make sure anyone want to add new predicate will not break your design here.

Comment (resouer) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: And as Bobby suggest, a test verify is needed to ensure it.

**Comment by resouer** [comment]
@yastij I think you can discuss this during sig-scheduling meeting to see if it's a proper candidate for 1.10 release

**Comment by yastij** [comment]
@resouer - I'll be there to discuss this + some other items

**Comment by yastij** [comment]
@resouer @bsalamat - updated tests & the CI is green. will refactor predicate names tomorrow and ping you for lgtm.

**Comment by yastij** [comment]
@bsalamat @resouer - PTAL for lgtm/approval

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 5]
```diff
@@ -79,6 +95,21 @@ const (
 // For example:
 // https://github.com/kubernetes/kubernetes/blob/36a218e/plugin/pkg/scheduler/factory/factory.go#L422
 
+// IMPORTANT: this list contains the ordering of the predicates, if you develop a new predicates
+// it is mandatory to add its name on this list.
+// otherwise it won't be processed, see generic_scheduler#podFitsOnNode()
+// the order is based on the restrictiveness & complexity of predicates
```
Comment (resouer) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: It seems to be `Otherwise it` `The order` , `see generic_scheduler#podFitsOnNode().` Please check again to avoid any typo nits

Comment (yastij) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: indeed, fixed

**Review by resouer — APPROVED** [review]
change lgtm overall, I will let @bsalamat do final pass since he has more context. 

Also, seems this need to be merged before #56926

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 6]
```diff
@@ -49,9 +49,25 @@ import (
 )
 
 const (
-	MatchInterPodAffinity = "MatchInterPodAffinity"
-	CheckVolumeBinding    = "CheckVolumeBinding"
-
+	MatchInterPodAffinity               = "MatchInterPodAffinity"
```
Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Please add "Pred" suffix to the first two items to be consistent with the rest.

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 7]
```diff
@@ -152,23 +152,23 @@ func defaultPredicates() sets.String {
...
-		factory.RegisterFitPredicate("CheckNodeMemoryPressure", predicates.CheckNodeMemoryPressurePredicate),
+		factory.RegisterFitPredicate(predicates.CheckNodeMemoryPressurePred, predicates.CheckNodeMemoryPressurePredicate),
 
 		// Fit is determined by node disk pressure condition.
-		factory.RegisterFitPredicate("CheckNodeDiskPressure", predicates.CheckNodeDiskPressurePredicate),
+		factory.RegisterFitPredicate(predicates.CheckNodeDiskPressurePred, predicates.CheckNodeDiskPressurePredicate),
 
-		// Fit is determined by node conditions: not ready, network unavailable or out of disk.
-		factory.RegisterMandatoryFitPredicate("CheckNodeCondition", predicates.CheckNodeConditionPredicate),
+		// Fit is determied by node condtions: not ready, network unavailable and out of disk.
```
Comment (bsalamat) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: accidental edits?

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 8]
```diff
@@ -79,6 +95,21 @@ const (
 // For example:
 // https://github.com/kubernetes/kubernetes/blob/36a218e/plugin/pkg/scheduler/factory/factory.go#L422
 
+// IMPORTANT: this list contains the ordering of the predicates, if you develop a new predicates
```
Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: s/predicates/predicate/

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 9]
```diff
@@ -79,6 +95,21 @@ const (
...
+// it is mandatory to add its name on this list.
```
Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: s/on/to/

**Review by bsalamat — COMMENTED** [review]
Thanks, @yastij! Looks good. Just minor comments.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 10]
```diff
@@ -93,6 +124,14 @@ type CachedPersistentVolumeInfo struct {
 	corelisters.PersistentVolumeLister
 }
 
+func GetPredicatesOrdering() []string {
```
Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Providing getters and setters is not common in our code, but you can add them if you like. However, getters shouldn't have "Get" in their names in Go conventions. You can rename this to `PredicatesOrdering`.

File: plugin/pkg/scheduler/scheduler_test.go [hunk 11]
```diff
@@ -43,6 +43,8 @@ import (
 	"k8s.io/kubernetes/plugin/pkg/scheduler/volumebinder"
 )
 
+var order = []string{"VolumeBindingChecker"}
```
Comment (bsalamat) [inline plugin/pkg/scheduler/scheduler_test.go]: This doesn't need to be file-scoped. Please put this inside the test that uses it and use `predicates.CheckVolumeBindingPred` instead of the literal string. Also, please add at least one more predicate to the list. I think `GeneralPredicates` would be a good choice.

**Comment by bsalamat** [comment]
/lgtm

Thanks, @yastij!

**Comment by resouer** [comment]
The test failure should be related, I will help to look at it later today

**Comment by yastij** [comment]
@resouer - can you undo the lgtm to avoid spam by the bot ?

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 12]
```diff
@@ -79,6 +95,21 @@ const (
...
+// Design doc: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/predicates-ordering.md
+var (
+	predicatesOrdering = []string{CheckNodeConditionPred,
+		GeneralPred, HostNamePred, PodFitsHostPortsPred,
+		MatchNodeSelectorPred, PodFitsResourcesPred, NoDiskConflictPred,
+		PodToleratesNodeTaintsPred, PodToleratesNodeNoExecuteTaintsPred, CheckNodeLabelPresencePred,
+		checkServiceAffinityPred, MaxEBSVolumeCountPred, MaxGCEPDVolumeCountPred,
+		MaxAzureDiskVolumeCountPred, CheckVolumeBindingPred, NoVolumeZoneConflictPred,
+		CheckNodeMemoryPressurePred, CheckNodeDiskPressurePred, MatchInterPodAffinityPred}
+)
```
Comment (resouer) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: TODO: We may need to work out some sorting mechanism in the future. So developers only need to register their predicate and it will be placed to right position. Then we don't need to maintain this hard coding list.

**Comment by yastij** [comment]
@resouer - the only way I can think of is to expose ordering through policy file, otherwise we have to hardcode ordering in our code. But as said @bsalamat we might want to wait and get feedback before doing it

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 12 above)
Comment (bsalamat) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: I doubt if we can have an intelligent mechanism that orders predicates for us. We try to run faster, more restrictive predicates first, so that the overall execution time of the predicates drop.

**Comment by bsalamat** [comment]
> @resouer - the only way I can think of is to expose ordering through policy file, otherwise we have to hardcode ordering in our code. But as said @bsalamat we might want to wait and get feedback before doing it

Making the ordering configurable is certainly possible, but as we had discussed in the past, it may have limited usage. We should wait and see if there is any request to add the feature.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 12 above)
Comment (resouer) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: I see, it's reasonable.

**Comment by resouer** [comment]
Slack talked with @yastij , the CI failure has been captured will be fixed soon, really appreciate his effort!

**Comment by yastij** [comment]
@resouer @bsalamat - CI all green, ready to ship. Thanks everyone !

**Comment by dims** [comment]
reapplying lgtm

/lgtm
